### PR TITLE
KAFKA-5597: Autogenerate producer sender metrics.

### PR DESCRIPTION
**Pull request by wushujames**
Subtask of https://issues.apache.org/jira/browse/KAFKA-3480

The changes are very similar to what was done for the consumer in https://issues.apache.org/jira/browse/KAFKA-5191 (pull request https://github.com/apache/kafka/pull/2993)

A screenshot of the docs are here:
producer metrics docs

### CODE REVIEW

**Comment by wushujames** [comment]
Pinging @guozhangwang for review, since he reviewed my previous metrics PR.

File: checkstyle/suppressions.xml [hunk 1]
```diff
@@ -32,6 +32,8 @@
     <suppress checks="ParameterNumber"
               files="Fetcher.java"/>
     <suppress checks="ParameterNumber"
+              files="Sender.java"/>
```
Comment (wushujames) [inline checkstyle/suppressions.xml:35]: Is this okay? I added a new parameter to the Sender class, which pushed it over the checkstyle limit.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        this.recordsPerRequestAvg = new MetricNameTemplate("records-per-request-avg", metricGrpName, "The average number of records per request.", tags);
+        this.recordRetryRate = new MetricNameTemplate("record-retry-rate", metricGrpName, "The average per-second number of retried record sends", tags);
+        this.recordErrorRate = new MetricNameTemplate("record-error-rate", metricGrpName, "The average per-second number of record sends that resulted in errors", tags);
+        this.recordSizeMax = new MetricNameTemplate("record-size-max", metricGrpName, "The maximum record size", tags);
+        this.recordSizeAvg = new MetricNameTemplate("record-size-avg", metricGrpName, "The average record size", tags);
+        this.requestsInFlight = new MetricNameTemplate("requests-in-flight", metricGrpName, "The current number of in-flight requests awaiting a response.", tags);
+        this.metadataAge = new MetricNameTemplate("metadata-age", metricGrpName, "The age in seconds of the current producer metadata being used.", tags);
+        this.batchSplitRate = new MetricNameTemplate("batch-split-rate", metricGrpName, "The average number of batch splits per second", tags);
```
Comment (wushujames) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java]: I changed the description of this metric to be more readable. The previous one wasn't even in the docs, but in the code it said "The rate of record batch split"

**Comment by wushujames** [comment]
Why didn't this trigger automatic builds to run all the tests?

**Comment by wushujames** [comment]
Ah, it seems like my comment just now triggered the builds.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6406/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6391/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by wushujames** [comment]
Pinging @guozhangwang. I know you are running the 1.0.0 release, but do you have time to review this?

**Comment by guozhangwang** [comment]
@wushujames Sorry for not replying earlier. I will try to review the PR next week; if I got pulled away, @ijuma may chime in and review it later.

**Comment by wushujames** [comment]
I just merged in the latest changes from trunk into this PR. @guozhangwang or @ijuma, will one of you be able to review this?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6839/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6825/
Test PASSed (JDK 8 and Scala 2.12).

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java [hunk 3]
```diff
@@ -674,22 +675,19 @@ public void wakeup() {
         this.client.wakeup();
     }
 
-    public static Sensor throttleTimeSensor(Metrics metrics) {
-        String metricGrpName = SenderMetrics.METRIC_GROUP_NAME;
+    public static Sensor throttleTimeSensor(Metrics metrics, SenderMetricsRegistry metricsRegistry) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:680]: This is not a comment for the PR, but just a question for my own knowledge to @ijuma : this throttle time sensor is added to the network client layer, but we make it inside `Sender`. Why is this design?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java [hunk 4]
```diff
@@ -231,7 +230,8 @@ public boolean matches(AbstractRequest body) {
     @Test
     public void testQuotaMetrics() throws Exception {
         MockSelector selector = new MockSelector(time);
-        Sensor throttleTimeSensor = Sender.throttleTimeSensor(metrics);
+        SenderMetricsRegistry metricsRegistry = new SenderMetricsRegistry(metrics.config().tags().keySet());
```
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java]: Can we make a `SenderMetricsRegistry metricsRegistry` a private field of the class that takes the private `metrics` object, and reuse it in all the test functions?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+        this.produceThrottleTimeAvg = new MetricNameTemplate("produce-throttle-time-avg", metricGrpName, "The average time in ms a request was throttled by a broker", tags);
+        this.produceThrottleTimeMax = new MetricNameTemplate("produce-throttle-time-max", metricGrpName, "The maximum time in ms a request was throttled by a broker", tags);
+
+        /* ***** Topic level *****/
+        String topicMetricGrpName = "producer-topic-metrics";
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java]: nit: we can make it final static. Same for `metricGrpName` above.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java (hunk 2 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java]: Good catch!

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+    public SenderMetricsRegistry(Set<String> tags) {
+
+        /* ***** Client level *****/
+        String metricGrpName = "producer-metrics";
+        
+        this.batchSizeAvg = new MetricNameTemplate("batch-size-avg", metricGrpName, "The average number of bytes sent per partition per-request.", tags);
+        this.batchSizeMax = new MetricNameTemplate("batch-size-max", metricGrpName, "The max number of bytes sent per partition per-request.", tags);
+        this.compressionRateAvg = new MetricNameTemplate("compression-rate-avg", metricGrpName, "The average compression rate of record batches.", tags);
+        this.recordQueueTimeAvg = new MetricNameTemplate("record-queue-time-avg", metricGrpName, "The average time in ms record batches spent in the record accumulator.", tags);
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java]: nit: maybe rephrase `record accumulator` to `send buffer` to be more understandable, as record accumulator is an internal classname?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java [hunk 7]
```diff
@@ -700,68 +698,69 @@ public static Sensor throttleTimeSensor(Metrics metrics) {
...
-            m = metrics.metricName("record-retry-rate", metricGrpName, "The average per-second number of retried record sends");
+            m = metrics.metricInstance(metricsRegistry.recordRetryRate);
             this.retrySensor.add(m, new Rate());
 
             this.errorSensor = metrics.sensor("errors");
-            m = metrics.metricName("record-error-rate", metricGrpName, "The average per-second number of record sends that resulted in errors");
+            m = metrics.metricInstance(metricsRegistry.recordErrorRate);
             this.errorSensor.add(m, new Rate());
 
             this.maxRecordSizeSensor = metrics.sensor("record-size-max");
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:745]: Not a comment for this PR: the sensor name seems wrong, as it should be `record-size`, but changing it would be a breaking change. cc @ijuma .

File: checkstyle/suppressions.xml (hunk 1 above)
Comment (guozhangwang) [reply checkstyle/suppressions.xml:35]: Fine with me.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetrics.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.Set;
+
+import org.apache.kafka.common.MetricNameTemplate;
+import org.apache.kafka.common.metrics.Metrics;
+
+public class ProducerMetrics {
+
+    public SenderMetricsRegistry senderMetrics;
+
+    public ProducerMetrics(Set<String> tags, String metricGrpPrefix) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetrics.java:31]: The second parameter is not needed?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetrics.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    private List<MetricNameTemplate> getAllTemplates() {
+        List<MetricNameTemplate> l = new ArrayList<>();
+        l.addAll(this.senderMetrics.getAllTemplates());
+        return l;
+    }
+
+    public static void main(String[] args) {
+        Set<String> tags = new HashSet<>();
+        tags.add("client-id");
+        ProducerMetrics metrics = new ProducerMetrics(tags, "producer");
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetrics.java:44]: Ditto above, `producer` string seems not needed.

**Review by guozhangwang — COMMENTED** [review]
Left some comments, thanks @wushujames . A few non-PR related questions cc @ijuma .

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java (hunk 3 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:680]: I agree that this is a bit weird and I asked the same question. If I remember correctly, it's done this way to record the values in a generic way, but the names are only known by the callers (depending on whether it's a producer or consumer). @rajinisivaram can elaborate.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java (hunk 7 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:745]: Good catch. Is it a breaking change? The metric name is correct. Does the sensor name appear in the JMX name?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java (hunk 3 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:680]: Yes, the sensors are created in `Sender/Fetcher` to avoid knowledge of the different names to the network layer. Recording is done in the network layer since `Sender/Fetcher` don't see all responses (now that any response may be throttled) and to use common logic.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java (hunk 7 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:745]: Sensor names don't appear in JMX.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetrics.java (hunk 8 above)
Comment (wushujames) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetrics.java:31]: You're right, it's not used in this checkin. But it'll actually get used once I convert more metrics over to this technique. In particular, the SelectorMetrics are used in both the producer and consumer. Stuff like `request-rate` and `connection-close-rate`. When used in the producer, the MBean is something like `kafka.producer:type=producer-metrics` and when used in the consumer, the MBean is something like `kafka.consumer:type=consumer-metrics`. So this 2nd parameter is passed in and used to contruct the MBean name.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetrics.java (hunk 9 above)
Comment (wushujames) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetrics.java:44]: Yup. Not used in this checkin, but will be in the future. See above for why.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java (hunk 4 above)
Comment (wushujames) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java]: Unfortunately, the answer is both yes and no. 

Yes, in this particular case (in `testQuotaMetrics()`)

In the rest of the places, no, not without touching more lines of code. The reason is, the setup() method sets up the private `metrics` object. However, most of the tests actually define _their own_ metrics object (like, see Line 237 below). And the two aren't interchangeable, because they were set up with different tags. And, similarly, the SenderMetricsRegistry() has to be set up using the same tags as the corresponding `metrics` object, so we can make the SenderMetricsRegistry() interchangeable as well.

That said, I will make the setup() method also set up a `senderMetricsRegistry` object. I will switch testQuotaMetrics() to use it and then in the future, if/when this file gets cleaned up to all use the same `metrics` object, then we will make them all use the same `senderMetricsRegistry` object.

I could switch all the tests to use the _same_ metrics object, but I didn't want to do a lot of refactoring in this PR. I wanted to keep this PR focused on adding the Sender metrics.

**Comment by wushujames** [comment]
I merged in the latest trunk, and fixed some of the small requested changes. I also responded to everyone's questions in the PR. Can someone take another look?

**Comment by guozhangwang** [comment]
Merged to trunk after fixing the naming bug of `record-size`. Thanks @wushujames !!
